runtime-rs: Support cgroup v2 on host

yaoyinnan opened the pull request:
Support cgroup v2 in runtime-rs, reference to https://github.com/kata-containers/kata-containers/pull/4397, https://github.com/kata-containers/kata-containers/issues/4886 and https://github.com/kata-containers/kata-containers/pull/5153.

The cgroups-rs repository currently does not support setting threaded mode in cgroup v2. Support for threaded mode has been added in https://github.com/kata-containers/cgroups-rs/pull/91. Currently, the version of cgroups-rs in the dependency is replaced by branch https://github.com/kata-containers/cgroups-rs/pull/91.

Fixes: #5109

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

liubin commented:
I added a `wip` label to wait the https://github.com/kata-containers/cgroups-rs/pull/91 to be merged.

studychao commented:
Thanks ! Yinnan! Will take a look later

yaoyinnan commented:
/test

yaoyinnan commented:
/test

yaoyinnan commented on src/runtime-rs/crates/resource/Cargo.toml:
```diff
@@ -12,7 +12,8 @@ test-utils = { path = "../../../libs/test-utils" }
 bitflags = "1.2.1"
-cgroups-rs = "0.2.9"
+#cgroups-rs = "0.2.9"
+cgroups-rs = { git = "https://github.com/yaoyinnan/cgroups-rs.git", branch = "90/ftr/threaded-mode" }
```
Yes, but if there is something to be modified later, two PRs may be required to be modified at the same time.

yaoyinnan commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -33,22 +34,34 @@ pub struct CgroupConfig {
+        let is_threaded_mode = is_v2 && !toml_config.runtime.sandbox_cgroup_only;
+
+        let sandbox_path = if is_threaded_mode {
+            overhead_path = format!("{}/overhead", path);
```
I added comments.

yaoyinnan commented:
/test

YushuoEdge commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -33,22 +33,49 @@ pub struct CgroupConfig {
+        // the same cgroup in domain threaded mode.
+        // In this way, vCPU threads and VMM processes can be separated into two cgroups.
+        // For details, please refer to host cgroups design document
+        // https://github.com/kata-containers/kata-containers/blob/main/docs/design/host-cgroups.md.
```
Do not use main/master in GitHub URL, use tag/commit instead (?

yaoyinnan replied:
Thanks for your review. @YushuoEdge

Is it to prevent the link from being incorrectly accessed due to changes in files on the main branch?

After https://github.com/kata-containers/kata-containers/pull/5153 is merged, I will write the file link of the docs commit in the comments here.

YushuoEdge replied:
Yes, thanks!

YushuoEdge reviewed (commented):
Thanks @yaoyinnan ! A small question/problem.

yaoyinnan commented:
/test

Tim-Zhang commented on src/runtime-rs/crates/resource/src/cgroups/utils.rs:
```diff
@@ -14,3 +25,20 @@
+    };
+
+    let cgroup_path_dir = cgroup_path.parent().unwrap().to_string_lossy().to_string();
+    let cgroup_path_name = format!(
```
cgroup_path_name could be invalid when cgroup_path = Path::new(DEFAULT_RESOURCE_CONTROLLER_ID)

yaoyinnan replied:
Thanks for your review, I fixed it.

Tim-Zhang commented:
Does this need a rework based on cgroup-rs@0.3.0? @yaoyinnan

yaoyinnan commented:
> Does this need a rework based on cgroup-rs@0.3.0? @yaoyinnan

@Tim-Zhang It has been reworked to be based on cgroup-rs@0.3.0.

yaoyinnan commented:
/test

yaoyinnan commented:
/test

Tim-Zhang commented on src/runtime-rs/Cargo.lock:
```diff
@@ -412,6 +412,18 @@ dependencies = [
 
+[[package]]
+name = "cgroups-rs"
+version = "0.2.11"
```
But this version is not right here if you have reworked it already

yaoyinnan replied:
updated

Tim-Zhang commented on src/runtime-rs/crates/resource/src/cgroups/utils.rs:
```diff
@@ -14,3 +25,19 @@
+    let cgroup_path_name = format!(
+        "{}_{}",
+        CGROUP_KATA_PREFIX,
+        cgroup_path.file_name().unwrap().to_str().unwrap()
```
if cgroup_path is '/', this sentence would be panic

yaoyinnan replied:
fixed, and I added some UT.

yaoyinnan commented:
/test

yaoyinnan commented:
/test

BbolroC commented:
Regarding the CI errors for `[s390x|ppc64le]-snap`, it has been resolved by https://github.com/kata-containers/kata-containers/pull/6212. Please rebase this onto main.

yaoyinnan commented:
/test

yaoyinnan commented:
/test

liubin commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -133,13 +216,90 @@ impl CgroupsResource {
+                .context("delete cgroup manager")?;
+
+            // Delete overhead cgroup manager.
+            if let Some(overhead) = self.overhead_cgroup_manager.as_ref() {
```
Can line 276 and 259 be combined into one `if let ...`?

yaoyinnan replied:
merged together

yaoyinnan commented:
/test

yaoyinnan commented:
/test-s390x

yaoyinnan commented:
/test-s390x

yaoyinnan commented:
@bergwolf @fidencio @liubin This PR has basically been completed. Can you review and push to approve this PR for me?

yaoyinnan commented:
/test

justxuewei commented on src/runtime-rs/crates/resource/src/cgroups/utils.rs:
```diff
@@ -14,3 +25,61 @@
+            },
+        ];
+
+        for (_i, t) in tests.iter().enumerate() {
```
It is not necessary to have an `enumerate()`.

yaoyinnan replied:
deleted

justxuewei replied:
After moved the `enumerate()`, then the for loop should be `for t in tests.iter()` instead of `for (_i, t) in tests.iter()`.

yaoyinnan replied:
fixed

justxuewei commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -133,13 +216,86 @@ impl CgroupsResource {
+            // Move threads to parent sandbox cgroup manager from cgroup manager that thread mode.
+            for cg_tid in self.cgroup_manager.tasks() {
+                // For now, we can't guarantee that the thread in cgroup_manager does still
+                // exist. Once it exit, we should ignor that error returned by remove_task
```
Nit: ignor -> ignore

yaoyinnan replied:
thanks, fixed typo

justxuewei reviewed (commented):
Thanks! A few comments here.

quanweiZhou commented on src/runtime-rs/crates/resource/src/cgroups/utils.rs:
```diff
@@ -14,3 +25,61 @@
     format!("kata_overhead/{}", path.trim_start_matches('/'))
 }
+
+pub(crate) fn rename_cgroup_path(path: &str) -> String {
```
1. The `rename_cgroup_path()` seems like a generic function name, and the implementation here is to add the prefix kata for cgroup path, suggest rename like `add_kata_prefix_for_cgroup_path()`.
2. Suggest adding a comment for this function.

yaoyinnan replied:
renamed and added a comment

quanweiZhou reviewed (commented):
Thanks @yaoyinnan, I was wondering if it is possible to abstract the logic of cgroup operation through trait, and then implement the logic of v1/v2.
```
trait CgroupOperator {
  fn delete_cgroup();
  ...
}

impl CgroupOperator for CgroupOperatorV1 {}
impl CgroupOperator for CgroupOperatorV1 {}
```

gkurz commented:
Hi @yaoyinnan ! I got summoned on this PR by @Apokleos.

I can certainly help to get this reviewed but it will need some rework. First, please drop the first and second patches which seem to be totally unrelated to cgroup v2 support. Then you should understand that the greatest speed reducer for this PR is the monolithic third patch. It requires reviewers to devote a huge chunk of consecutive time to be able to review without context switches... for maintainers, which often need to work on several topics at the same time, this acts as a powerful repellent.

I am also pretty sure that there is more to say in the commit message than just `Support cgroup v2 in runtime-rs. Using thread mode to manager sandbox and overhead in cgroup v2.` .

If I am to invest time in this PR, I will ask to break down into multiple patches with detailed commit messages.

yaoyinnan commented:
> Hi @yaoyinnan ! I got summoned on this PR by @Apokleos.
> 
> I can certainly help to get this reviewed but it will need some rework. First, please drop the first and second patches which seem to be totally unrelated to cgroup v2 support. Then you should understand that the greatest speed reducer for this PR is the monolithic third patch. It requires reviewers to devote a huge chunk of consecutive time to be able to review without context switches... for maintainers, which often need to work on several topics at the same time, this acts as a powerful repellent.
> 
> I am also pretty sure that there is more to say in the commit message than just `Support cgroup v2 in runtime-rs. Using thread mode to manager sandbox and overhead in cgroup v2.` .
> 
> If I am to invest time in this PR, I will ask to break down into multiple patches with detailed commit messages.

Hi, thank you very much for your suggestions. I removed the two irrelevant commits. I saw that other people on Slack mentioned that they were dealing with the deref problem.

I split the original commit into three commits:

1. Add support for cgroup v2 in runtime-rs.
2. Depending on the cgroup version differences, delete the cgroup in different ways.
3. Add kata prefix to cgroup path.

yaoyinnan commented:
@gkurz In order to facilitate your understanding of why this PR uses threaded mode to handle sandbox and overhead cgroups, I have made the following explanation:

In the current cgroupv2, because the process mode cannot separate processes and threads into different cgroups, the vmm process is incorrectly added to the sandbox cgroup. The sandbox and overhead are placed in the same cgroup through thread mode to realize that the vCPU thread is placed in the sandbox and other processes are placed in the overhead.

![image](https://github.com/kata-containers/kata-containers/assets/35447132/30c7ad0e-d4bc-4a7f-8726-25542e6a3b38)

justxuewei commented on src/runtime-rs/crates/resource/src/cgroups/utils.rs:
```diff
@@ -4,6 +4,17 @@
 
+use std::path::Path;
+
+// prepend a kata specific string to oci cgroup path to
```
// prepend -> // Prepend

yaoyinnan replied:
Resolved

justxuewei commented on src/runtime-rs/crates/resource/src/cgroups/utils.rs:
```diff
@@ -4,6 +4,17 @@
+// from grabbing the stats data.
+const CGROUP_KATA_PREFIX: &str = "kata";
+
+// DEFAULT_RESOURCE_CONTROLLER_ID runtime-determined location in the cgroups hierarchy.
+const DEFAULT_RESOURCE_CONTROLLER_ID: &str = "vc";
```
I have two questions here:

1. What does `vc` stand for? VirtContainer?
2. Bear with me in the case I'm wrong. The `DEFAULT_RESOURCE_CONTROLLER_ID` is to avoid that the path is empty or "/". But why is the `kata` prefix not needed?

yaoyinnan replied:
1. The `vc` stands for VirtContainer.
2. Sorry, I missed it, I added it.

justxuewei commented on src/runtime-rs/crates/resource/src/cgroups/cgroup_persist.rs:
```diff
@@ -10,4 +10,5 @@ pub struct CgroupState {
     pub path: Option<String>,
     pub overhead_path: Option<String>,
     pub sandbox_cgroup_only: bool,
+    pub is_threaded_mode: bool,
```
Use a more concise name: `is_threaded_mode` -> `threaded_mode`

yaoyinnan replied:
replaced

justxuewei commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -35,30 +37,59 @@ pub struct CgroupArgs {
             .map(|linux| linux.cgroups_path.trim_start_matches('/').to_string())
             .unwrap_or_default();
 
+        path = utils::add_kata_prefix_for_cgroup_path(&path);
+        overhead_path = utils::add_kata_prefix_for_cgroup_path(&overhead_path);
+
+        // In cgroup v2, the sandbox and overhead cgroup in threaded mode are placed under
+        // the same cgroup in domain threaded mode.
+        // In this way, vCPU threads and VMM processes can be separated into two cgroups.
+        // For details, please refer to host cgroups design document
+        // https://github.com/kata-containers/kata-containers/blob/main/docs/design/host-cgroups.md.
+        let is_v2 = hierarchies::is_cgroup2_unified_mode();
+        let is_threaded_mode = is_v2 && !toml_config.runtime.sandbox_cgroup_only;
+        let specified_controllers = if is_threaded_mode {
+            Some(vec![String::from("cpuset"), String::from("cpu")])
+        } else {
+            None
+        };
+
+        let sandbox_path = if is_threaded_mode {
+            overhead_path = format!("{}/overhead", path);
+            format!("{}/sandbox", path)
+        } else {
+            path
+        };
```
The code might be optimized as following: mut removal, reduction of assigned value.

```
// remove mut
let overhead_path = utils::gen_overhead_path(sid);
let spec = load_oci_spec()?;
// remove mut
let path = utils::add_kata_prefix_for_cgroup_path(&spec
// ...

// remove this
// path = utils::add_kata_prefix_for_cgroup_path(&path);
// remove this
// overhead_path = utils::add_kata_prefix_for_cgroup_path(&overhead_path);

// ...

let (path, overhead_path) = if is_threaded_mode {
    (format!("{}/sandbox", path), format!("{}/overhead", path))
} else {
    (path, utils::add_kata_prefix_for_cgroup_path(&utils::gen_overhead_path(sid)))
}
```

yaoyinnan replied:
I followed your suggestion, and wrapped two functions to process the path.

justxuewei commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -111,6 +152,46 @@ impl CgroupsResource {
+                Err(anyhow!(
+                    "unable to match specified controller as allowed by thread mode."
+                ))
+            }
```
```
let sc = config.specified_controllers.clone().ok_or(anyhow!("error message"))?
```

yaoyinnan replied:
Yes, this way is better

justxuewei commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -111,6 +152,46 @@ impl CgroupsResource {
+        } else {
+            let hier = hierarchies::auto();
+            if config.is_threaded_mode {
+                if let Some(sc) = config.specified_controllers.clone() {
+                    Ok(Some(
+                        CgroupBuilder::new(&config.overhead_path)
+                            .set_specified_controllers(sc)
+                            .build(hier)?,
+                    ))
+                } else {
+                    Err(anyhow!(
+                        "unable to match specified controller as allowed by thread mode."
+                    ))
```
Ditto.

yaoyinnan replied:
Ditto.

justxuewei reviewed (commented):
Thanks @yaoyinnan! Lgtm overall, but here are some comments.

justxuewei commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -354,6 +327,147 @@ impl CgroupsResource {
+    fn do_delete_cgroups(&self, resource: &CgroupsResource) -> Result<()>;
+}
+
+struct CgroupOperatorV1;
+struct CgroupOperatorV2;
```
Ignore this. I saw above comments...

yaoyinnan commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -354,6 +327,147 @@ impl CgroupsResource {
+
+        // Move procs to root cgroup manager from cgroup manager.
+        for cg_pid in parent_cgroup_manager.procs() {
+            parent_cgroup_manager.remove_task_by_tgid(cg_pid)?;
```
added

yaoyinnan commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -354,6 +327,147 @@ impl CgroupsResource {
+                        None => return Err(error.into()),
+                    },
+                    None => return Err(error.into()),
+                }
```
I wrapped this into one method and also wrapped some other methods of moving threads.

yaoyinnan commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -354,6 +327,147 @@ impl CgroupsResource {
+    fn delete_non_threaded_cgroups(&self, resource: &CgroupsResource) -> Result<()> {
+        for cg_pid in resource.cgroup_manager.procs() {
+            resource.cgroup_manager.remove_task_by_tgid(cg_pid)?;
+        }
```
> Forget to deal with `OS_ERROR_NO_SUCH_PROCESS` error?

Use the `ignore_cgroup_move_error` method to handle `OS_ERROR_NO_SUCH_PROCESS` error.

yaoyinnan replied:
> Do we need to move children process to parent in non-threaded cgroup?

For cgroup v2, the process cannot be moved to the `k8s.io` cgroup (parent cgroup), but can only be moved to `root` cgroup.

yaoyinnan commented:
@justxuewei Thanks for your review. I made modifications based on your comments, please review again.

justxuewei reviewed (commented):
Great work! Last comments here ;)

justxuewei commented:
Btw, the title of this PR is able to be changed to "runtime-rs: support cgroup v2 on host".

justxuewei commented:
Plus, please rebase atop of the main branch.

yaoyinnan commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -74,26 +80,92 @@ impl CgroupsResource {
         } else {
-            None
+            CgroupBuilder::new(&config.path).build(hier)?
         };
```
modified

yaoyinnan commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -74,26 +80,92 @@ impl CgroupsResource {
+                    .build(hier)?
+            } else {
+                CgroupBuilder::new(&config.overhead_path).build(hier)?
+            };
```
modified

yaoyinnan commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -273,6 +246,133 @@ impl CgroupsResource {
+            self.delete_threaded_cgroup_manager(resource)?;
+        } else {
+            self.delete_domain_cgroup_manager(resource)?;
+        }
```
Yes, it's better handled this way, I also added the description

yaoyinnan commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -273,6 +246,133 @@ impl CgroupsResource {
+    fn do_delete_cgroups(&self, resource: &CgroupsResource) -> Result<()>;
+
+    fn move_tasks_to_root(&self, cgroup_manager: &Cgroup) -> Result<()> {
+        for cg_tid in cgroup_manager.tasks() {
```
Here is the move thread, I think cg_tid should be used, indicating thread id

yaoyinnan commented on src/runtime-rs/crates/resource/src/cgroups/mod.rs:
```diff
@@ -273,6 +246,133 @@ impl CgroupsResource {
+    }
+
+    fn move_tasks_to_parent(&self, cgroup_manager: &Cgroup) -> Result<()> {
+        for cg_tid in cgroup_manager.tasks() {
```
Here is the move thread, I think cg_tid should be used, indicating thread id

yaoyinnan commented:
@justxuewei Thank you for reviewing it again. I have modified and rebased.

justxuewei reviewed (approved):
Lgtm, thanks!
